Enable basic function calls from GPU kernels

daviditen opened the pull request:
Turn on simple function calls in the GPU eligibility checker function.

Add a new primitive to add to functions called from within GPU kernels to
enable codegen for both GPU and CPU. Codegen for both GPU and CPU
functions that are called from a GPU kernel or from functions that were
called from other functions called from the kernel.

Add errors for accessing variables declared outside of the function for
functions called from the GPU. I think we'd like to allow this eventually,
but for the time being it is an error.

Add new tests for calling functions from functions, and testing the error about
accessing outer variables. Add .futures for recursive function calls and
calling 'halt' which currently each silently cause the code to run on the CPU.

stonea reviewed (commented):
Looks good.

Are `FLAG_GPU_CODEGEN` and `FLAG_GPU_AND_CPU_CODEGEN` mutually exclusive?

Maybe rename `FLAG_GPU_CODEGEN` to `FLAG_GPU_ONLY_CODEGEN` and add a strategically placed assert to check that you don't have both flags set. `FLAG_GPU_ONLY_CODEGEN` is kind of a mouthful but I can't think of a shorter way
to express it.

--

Does one of your tests cover `FLAG_GPU_AND_CPU_CODEGEN`?  Just a test that would
call `foo()` both from within and outside of a kernel.

Could you also add a test that has a function that would have both GPU and CPU
codegen and could potentially do a kernel launch itself. Something like:

```chpl
proc foo() {
  // This forall loop should generate a kernel launch if not already on GPU:
  forall i in 0.n {
    // Do something
  }
}

forall i in 0..n {
  A[i] = foo();
}

foo();
```

daviditen commented:
> Are FLAG_GPU_CODEGEN and FLAG_GPU_AND_CPU_CODEGEN mutually exclusive?

No, I don't think they should be. FLAG_GPU_CODEGEN means something like, "generate a kernel to launch from the host onto the device". FLAG_GPU_AND_CPU_CODEGEN means generate a version callable from the kernel and a version callable from the host. I think there's a potential to want all three on a function.

> Does one of your tests cover FLAG_GPU_AND_CPU_CODEGEN? Just a test that would
> call foo() both from within and outside of a kernel.

I modified a test to call a function from both the host and device.

> Could you also add a test that has a function that would have both GPU and CPU
> codegen and could potentially do a kernel launch itself. Something like ...:

I added a test like you suggested. The forall directly in the on statement doesn't run on the GPU, but the forall in the called function does.

stonea commented:
I see, ok the nuance here is the flag isn't just saying something about when we do CPU vs GPU codegen but also what the calling convention for the function is. So if a function is flagged `GPU_AND_CPU_CODEGEN` then it won't be launched (as a
kernel) from the CPU.

So if I'm understanding correctly and I wasn't afraid of being overly verbose the names of the flags would be something like:

  `FLAG_GPU_CODEGEN_AND_KERNEL_LAUNCHABLE`
  `FLAG_GPU_AND_CPU_CODEGEN`

Not saying we should update the flag name to that, but maybe just add a comment to or update the comment string for `GPU_CODEGEN` in `PragmaList.h` to say something like: "generate GPU code; mark calling convention for function to be performed as a kernel launch".

Thanks for adding the additional test cases BTW.  Do we have now tests the cover all four possible configurations of `GPU_CODEGEN` and `GPU_AND_CPU_CODEGEN`?

I guess we don't need to worry about `(false, false)` (nearly all our tests stress that) and `(true, false)` is covered by plenty of existing GPU tests. I think `(false, true)` is covered by `callTrivialFn`. That just leaves `(true, true)`, I guess this case doesn't show up yet as the only functions that get flagged with `FLAG_GPU_CODEGEN` are ones that we automatically extract from forall\foreach loops.

daviditen commented:
Thanks!

e-kayrakli commented on compiler/optimizations/gpuTransforms.cpp:
```diff
@@ -316,14 +316,46 @@ static  CallExpr* generateGPUCall(OutlineInfo& info, VarSymbol* numThreads) {
+      if (var->defPoint->parentSymbol != fn) {
+        if (!var->isParameter() && var != gVoid) {
+          USR_FATAL(se, "variable '%s' must be defined in the function it"
+                    " is used in for GPU usage", var->name);
```
Why do we error here? Can't we just silently ignore the loop where a call has outer variable access?
